articles/_dev/2024-openui-advent-12.md

```diff
@@ -0,0 +1,207 @@
+
+```css
+
+:root {
+  color: light-dark(var(--light), var(--dark));
+  background-color: light-dark(var(--light-bg), var(--dark-bg));
+}
+
+```
+
+[CSS Color Module Level 5の仕様](https://drafts.csswg.org/css-color-5/#light-dark)によると、`light-dark()`関数は`color-scheme`プロパティによってテーマを決めることができます。従来からこの機能を持っていたのは`<system-color>`でしたが、`light-dark()`関数の登場により、`color-scheme`プロパティのテーマに依存した色の変更が可能になりました。
+
+> System colors have the ability to react to the current used color-scheme value. The light-dark() function exposes the same capability to authors.
+
+`light-dark()`関数は、`color-scheme`がライトテーマか不明な場合は1番目の`<color>`値を、ダークテーマの場合は2番目の`<color>`値を適用します。
+
+### `color-scheme`プロパティ
+
+`color-scheme`プロパティを使用すると、**ブラウザが**設定したカラーテーマを、要素に反映することができるCSS関数です。
+
+```css
+/* OSのライトテーマのみをサポートする */
+color-scheme: light;
+/* OSのダークテーマのみをサポートする */
+color-scheme: dark;
+/* 両方サポートする */
+color-scheme: light dark;
+/* ライトテーマを強制する */
+color-scheme: only light;
+/* ダークテーマを強制する */
+color-scheme: only dark;
+/* ページデフォルトのカラーテーマを使用する */
+color-scheme: normal;
+```
+
+***
+
+このように、`prefers-color-scheme`はユーザが適用するカラーテーマであるのに対し、`color-scheme`はブラウザが適用するカラーテーマです。
+そして、`prefers-color-scheme`に依存する`@media(prefers-color-scheme: <light | dark>)`は、「ユーザのカラーテーマ」を反映するのに対し、`color-scheme`に依存する`light-dark()`は、「ブラウザのカラーテーマ」を反映します。
+
+### カラースキーマの計算方法
+
+`color-scheme`は**ブラウザが**適用する色を決めるCSSプロパティですが、どのようにして実際に適用される色は決まるのでしょうか？
+
+> To **determine the used color scheme** of an element:
+>
+> 1. If the user’s preferred color scheme, as indicated by the prefers-color-scheme media feature, is present among the listed color schemes, and is supported by the user agent, that’s the element’s used color scheme.
+>
+> 2. Otherwise, if the user has indicated an overriding preference for their chosen color scheme, and the only keyword is not present in color-scheme for the element, the user agent must override the color scheme with the user’s preferred color scheme. See § 2.3 Overriding the Color Scheme.
+>
+> 3. Otherwise, if the user agent supports at least one of the listed color schemes, the used color scheme is the first supported color scheme in the list.
+>
+> 4. Otherwise, the used color scheme is the browser default. (Same as normal.)
+> https://drafts.csswg.org/css-color-adjust/#color-scheme-prop
+
+つまり、以下の順番でどのような色が適用されるか決まります。
+
+1. `prefers-color-scheme`の色が`color-scheme`によってサポートされる場合：`prefers-color-scheme`の色が適用される
+2. そうでない場合で、`color-scheme`で「only」を使用せずに「light/dark」が適用されている場合：`prefers-color-scheme`の色が適用される
+3. そうでない場合、つまり`color-scheme`に「only」を使用して「light/dark」が適用されている場合：`color-scheme`の色が適用される
+4. 上記いずれでもない場合：[`color-scheme: normal;`](https://drafts.csswg.org/css-color-adjust-1/#valdef-color-scheme-normal)の色が適用される。ページデフォルトの色が[`<meta name="color-scheme" content="dark">`](https://html.spec.whatwg.org/multipage/semantics.html#meta-color-scheme)で指定されている場合はその色が適用され、指定されていない場合はページデフォルトの色（通常はライトテーマ）が適用される
+
+上記の適用順序を確認できるデモを作成しました。長いのでCopepenリンクのみ記載します。
+
+デモ：
+
+https://codepen.io/sakupi01/pen/MYgjvwy
```

Comment:
解釈がおかしそう

case 2: only を指定してないから、ブラウザからの override があると上書きされる。がない場合はコンテンツの宣言が light なので、通常は light になる。(「prefers-color-scheme に応じた色になる」は間違い)

case 6(??): meta タグがなさそう？単に color-scheme がないから light-dark() が発動してないだけっぽい?

Reply:
> case 6(??)

元々はcase2, 3のdarkバージョンも作ってて、その時case6までできて、あとで絞った時に番号書き換え忘れてました！

Reply:
@Jxck 

case 2、修正しましたが、本当はこれ思った通りに動作していなくて、askしようと思っていたところでした。

以下によると、onlyなしの場合は、ユーザ設定からの上書きが可能で、例えばOSの設定やブラウザの設定を変えると、そのカラーテーマになると思ったんですが、demoだとそういう挙動はしません。（`color-scheme: light;`にしたら絶対light）
仕様の解釈間違っていそうでしょうか、、、？

> 2. Otherwise, if the user has indicated an overriding preference for their chosen color scheme, and the only keyword is not present in color-scheme for the element, the user agent must override the color scheme with the user’s preferred color scheme. See § 2.3 Overriding the Color Scheme.
> 2. そうでない場合、つまり`color-scheme`で「only」を使用せずに「light/dark」が適用されている場合：ユーザの設定した色が上書き適用される

> meta タグがなさそう？

追加しました！

https://codepen.io/sakupi01/pen/MYgjvwy

Reply:
上にも貼ったけど、 Chrome の場合 Override にあたるのは今のところ Auto Dark Mode 。
https://developer.chrome.com/blog/auto-dark-theme

Devtools で有効にしてみると、 only のない Case 2 も dark になってる。

<img width="1461" alt="image" src="https://github.com/user-attachments/assets/65602368-853c-459e-9aad-874028e0f326" />

Reply:
@Jxck あ〜〜〜、なるほど！！

> Otherwise, if the user has indicated an overriding preference for their chosen color scheme, and the only keyword is not present in color-scheme for the element, the user agent must override the color scheme with the user’s preferred color scheme. See § 2.3 Overriding the Color Scheme.
そうでない場合、つまりcolor-schemeで「only」を使用せずに「light/dark」が適用されている場合：ユーザの設定した色が上書き適用される


ずっと「ユーザから`prefers-color-scheme`の上書きができるってこと」という思い込みで読んでたので、auto-dark modeの方も同じ結果なんじゃないかと思ってました。auto-dark modeだと、「color-schemeのテーマを上書きできる」んですね。そして、仕様が言ってたのは「ユーザから~`prefers-color-scheme`~`color-scheme`の上書きができる」という意味合いうだったんですね。
理解しました、ありがとうございます！